### PR TITLE
DHCP DUID, IAID configuration options

### PR DESCRIPTION
**Pull request by vinaykul**
This integrates changes in https://github.com/systemd/systemd/pull/2818 and https://github.com/systemd/systemd/pull/2857, and fixes the public header file issue in  https://github.com/systemd/systemd/pull/2818#issuecomment-199515053.

Tested using our VM upgrade scenario:
IAID is specified in 10-eth0.network file in the [Link] section.
1. DUID specified in /etc/systemd/networkd.conf
2. DUID specified in /etc/systemd/network/10-eth0.network
3. DUID in networkd.conf overridden by DUID specified in 10-eth0.network.

IPv6 address migration was successful in all these cases. IPv4 address migration happened if ClientIdentifier=duid, and would not migrate if it is 'mac'. 

If DUID and IAID are not specified, systemd generates them based on machine-id and ifname respectively.

/cc @teg @poettering @keszybz @zonque @thom311 @nfnty

### CODE REVIEW

File: man/networkd.conf.xml [hunk 1]
```diff
@@ -0,0 +1,114 @@
+<?xml version='1.0'?> <!--*- Mode: nxml; nxml-child-indent: 2; indent-tabs-mode: nil -*-->
+<!DOCTYPE refentry PUBLIC "-//OASIS//DTD DocBook XML V4.2//EN"
+  "http://www.oasis-open.org/docbook/xml/4.2/docbookx.dtd">
+
+<!--
+  This file is part of systemd.
+
+  Copyright 2014 Tom Gundersen
```
Comment (keszybz) [inline man/networkd.conf.xml]: Not that this has any practical consequences, but why not put your name here?

File: man/networkd.conf.xml [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    For e.g. DHCP Unique Identifier (DUID).</para>
+
+  </refsect1>
+
+  <xi:include href="standard-conf.xml" xpointer="main-conf" />
+
+  <refsect1>
+    <title>[DUID] Section Options</title>
+
+    <para>This section configures the DUID value used by the DHCP protocol. The DUID value
```
Comment (keszybz) [inline man/networkd.conf.xml]: Please expand "DUID" for people who don't yet know what it is and are reading the man page to find out.

File: man/networkd.conf.xml [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    specified here overrides the DUID that systemd-networkd generates using the machine-id
+    from the <filename>/etc/machine-id</filename> file.  To configure DUID per-network, see
+    <citerefentry><refentrytitle>systemd.network</refentrytitle><manvolnum>5</manvolnum>
+    </citerefentry>.</para>
+
+    <para>The configured DHCP DUID should conform to the specification in
```
Comment (keszybz) [inline man/networkd.conf.xml:84]: Could you add a short statement about the relationship between DUID and IAID?

File: src/libsystemd-network/sd-dhcp6-client.c [hunk 4]
```diff
@@ -180,41 +180,29 @@ static int client_ensure_duid(sd_dhcp6_client *client) {
         return dhcp_identifier_set_duid_en(&client->duid, &client->duid_len);
 }
 
-int sd_dhcp6_client_set_duid(
-                sd_dhcp6_client *client,
-                uint16_t type,
-                uint8_t *duid, size_t duid_len) {
+int sd_dhcp6_client_set_duid(sd_dhcp6_client *client, uint16_t duid_type,
+                             size_t duid_len, uint8_t *duid) {
```
Comment (keszybz) [inline src/libsystemd-network/sd-dhcp6-client.c]: Why this change?

File: src/network/networkd-conf.c [hunk 5]
```diff
@@ -0,0 +1,154 @@
+/*-*- Mode: C; c-basic-offset: 8; indent-tabs-mode: nil -*-*/
+
+/***
+  This file is part of systemd.
+
+  Copyright 2014 Tom Gundersen <teg@jklm.no>
```
Comment (keszybz) [inline src/network/networkd-conf.c]: Likewise.

File: src/network/networkd-conf.c [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+        int r;
+        long byte;
+        char *cbyte, *pnext;
+        const char *pduid = (const char *)rvalue;
```
Comment (keszybz) [inline src/network/networkd-conf.c]: Why the cast?

File: src/libsystemd-network/sd-dhcp6-client.c (hunk 4 above)
Comment (vinaykul) [inline src/libsystemd-network/sd-dhcp6-client.c]: I renamed the parameters for more clarity. Looking more closely, it does fall out of line with the systemd pattern  - f(data\* , size_t) vs. f(size_t , data *) .

I just have a habit of writing it the latter way. I'll fix it to read:
sd_dhcp6_client_set_duid(sd_dhcp6_client *client, uint16_t duid_type,
                                           uint8_t *duid, size_t duid_len)

File: src/systemd/sd-dhcp6-client.h [hunk 7]
```diff
@@ -85,8 +85,9 @@ int sd_dhcp6_client_set_index(sd_dhcp6_client *client, int interface_index);
 int sd_dhcp6_client_set_local_address(sd_dhcp6_client *client, const struct in6_addr *local_address);
 int sd_dhcp6_client_set_mac(sd_dhcp6_client *client, const uint8_t *addr,
                             size_t addr_len, uint16_t arp_type);
-int sd_dhcp6_client_set_duid(sd_dhcp6_client *client, uint16_t type, uint8_t *duid,
-                             size_t duid_len);
+int sd_dhcp6_client_set_duid(sd_dhcp6_client *client, uint16_t duid_type,
+                             size_t duid_len, uint8_t *duid);
+int sd_dhcp6_client_set_iaid(sd_dhcp6_client *client, uint32_t iaid);
```
Comment (keszybz) [inline src/systemd/sd-dhcp6-client.h:90]: Why this change?

**Comment by keszybz** [comment]
So... you didn't change the naming of the parameters in any way. But looking at your patch, I think the naming should beg discussed. In particular, why not simply call the option `DUID=` or `UniqueIdentifier=`?

**Comment by vinaykul** [comment]
@keszybz While the DUID that goes out on the wire is an array of bytes, it is composed of a few different pieces of information depending on the type of DUID used. Ref: https://tools.ietf.org/html/rfc3315#section-9.1 

My initial thoughts was to keep it simple and just specify DUIDValue=XX:XX:XX:XX per the discussion in issue https://github.com/systemd/systemd/issues/394. However, that is not the most user-friendly way to configure it, and @teg is looking for the ability to specify the type and the parts. 

There is value in making the configuration user-friendly, so I picked up the naming spec that Tom suggested from https://github.com/systemd/systemd/pull/2761#issuecomment-190835118 with agreement to implement 'raw' at this time, while allowing room for future extensibility.

Does that address your concerns?

File: src/systemd/sd-dhcp6-client.h (hunk 7 above)
Comment (vinaykul) [inline src/systemd/sd-dhcp6-client.h:90]: sd_dhcp6_client_set_iaid is a new API needed to configure IAID. I reversed the parameter order change in sd_dhcp6_client_set_duid and I'm testing the fix. I'll send an update shortly.

File: src/libsystemd-network/sd-dhcp-client.c [hunk 8]
```diff
@@ -298,6 +298,52 @@ int sd_dhcp_client_set_client_id(sd_dhcp_client *client, uint8_t type,
...
+                r = dhcp_validate_duid_len(client->client_id.type, duid_len);
+                if (r < 0)
+                        return r;
+                client->client_id.ns.duid.type = htobe16(duid_type);
+                memcpy(&client->client_id.ns.duid.raw.data, duid, duid_len);
+                duid_len += sizeof(client->client_id.ns.duid.type);
+        }
+
+        client->client_id_len = sizeof(client->client_id.type) + duid_len +
+                                sizeof(client->client_id.ns.iaid);
```
Comment (keszybz) [inline src/libsystemd-network/sd-dhcp-client.c:336]: Is sizeof(type) counted twice? Once at line 332, and once here?

File: src/libsystemd-network/network-internal.c [hunk 9]
```diff
@@ -335,6 +335,34 @@ int config_parse_hwaddr(const char *unit,
...
+                            void *userdata) {
+        uint32_t iaid_value;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        if ((r = safe_atou32(rvalue, &iaid_value)) < 0) {
```
Comment (keszybz) [inline src/libsystemd-network/network-internal.c]: The assignment should be on a separate line.

**Comment by keszybz** [comment]
What @teg proposed in https://github.com/systemd/systemd/issues/394#issuecomment-124225892 is in fact very close to what I suggested above. At least in #394 there is no discussion of a separate section. I don't understand why add a separate section to hold one or two options. 

Also note @poettering's comment https://github.com/systemd/systemd/issues/394#issuecomment-124228991: "add a logic so that `DUIDType=` is determined automatically". Right now there's only a single value possible, so `DUIDType=raw` is completely redundant. When and if other options are added, we can add `DUIDType=` to disambiguate.

To make things concrete, what about the following:

```
[Link]
DHCPUniqueIdentifier=
DHCPIdentityAssociationIdentifier=
```

This is simple and user friendly, and future extensibility is preserved.
Incidentally it also matches @teg's proposal (except that "DUID" is replaced by "DHCP" to avoid repeating "UID" more than once).

**Comment by vinaykul** [comment]
@keszybz: @teg suggested a separate [DUID] section in https://github.com/systemd/systemd/pull/2761#issuecomment-190835118 and his suggestion for a separate section makes sense given the different options we want to support going forward. He had already mentioned having such a structure for DUID in issue #394 in comment https://github.com/systemd/systemd/issues/394#issuecomment-124225892.

If we just had DHCPIdentityAssociationIdentifier= and let's say the user wants to specify DUID-LLT. They would need to know the hardware address or decide which hardware address to use, its size and  format and how to find it, know to represent time in seconds from UTC Jan 1 2000, know the hardware type, and construct this after essentially mastering RFC 3315 section 9.

Alternately, they could just say 'Type=link-layer-and-time' and we would (by default) pick up the hardware address of eth0 (or link with lowest ifindex) if MACAddress is not specified, pick up the current time if TimeStamp is not specified, and construct DUID based on the hardware type of the default interface, fill in the code in network order. Extending on this, we can allow the user to specify hardware-address, and let us fill in the time (default to current time), and yet allow the user to have DUID-LLT. This would not be possible without the Type field.

I think this is significantly more user-friendly, and this is the value I see in the spec that Tom proposed. It was significantly lot less work to just have raw format without this extensibility support but doing it this way has the above benefits. So having a separate section for it also makes sense. Thoughts? 

For the present time though since only Raw is supported, I can modify the code to presume if Type is not specified, Type=raw (default). Does that help simplify things?

File: src/libsystemd-network/sd-dhcp-client.c (hunk 8 above)
Comment (vinaykul) [inline src/libsystemd-network/sd-dhcp-client.c:336]: No.  duid_len only holds the size of raw DUID data that gets copied to client_id.ns.duid.raw.data.

client_id_len should hold the size of the entire client identifier that goes on the wire: i.e sizeof(client identifier type) + sizeof(iaid) + duid size ( = sizeof(ns.duid.type) + duid_len ). 

Per RFC 4361: https://tools.ietf.org/html/rfc4361#section-6.1, this adds up to 19 bytes for DUID-EN. 

screen shot 2016-03-25 at 1 31 27 pm

File: src/libsystemd-network/network-internal.c (hunk 9 above)
Comment (vinaykul) [inline src/libsystemd-network/network-internal.c]: OK. I'll fix this and update the code review change.

**Comment by vinaykul** [comment]
@keszybz Is it possible to come to an agreement with @teg on the interface we want to use and finalize on it. 

Your latest suggestion is close to the simple approach that I initially suggested on Feb 17th: https://github.com/systemd/systemd/issues/394#issuecomment-185489519. I revised my first patch based on suggestions by @poettering and @teg , and took up Tom's suggested [DUID] section in https://github.com/systemd/systemd/pull/2761

Now we seem to want to backtrack to the simpler approach again.

We made changes to our internal code that consumes this configuration interface based on Tom merging my change. Going back and forth this way is very disruptive to our testing team and their sign-off processes.

Could you please help drive this to an agreed upon interface? Both the simple raw DUID configuration interface, as well as the more configurable [DUID] section based interface works for us. We just need it finalized one way or another.

Thanks for the code-reviews and your help.

**Comment by keszybz** [comment]
OK. Looking at this again, the approach with a separate DUID section is fine. I'd have done this differently, but that's just bike-shedding, and everybody else seems to like current approach, so let's keep that.

But please rename `IAIDValue` to `IdentityIdentifier` (or `IAID`). In fact you refer to it as "IAID" in messages already, so it'll be more consistent to drop "Value".

Let's fix some final details in the docs before merging though. Also please squash the two commits into one and push again.

File: man/networkd.conf.xml [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+  <xi:include href="standard-conf.xml" xpointer="main-conf" />
+
+  <refsect1>
+    <title>[DUID] Section Options</title>
+
+    <para>This section configures the DHCP Unique Idendifier (DUID) value used by DHCP
+    protocol. DHCPv6 client protocol sends the DHCP Unique Identifier and the interface
+    Identity Association Identifier (IAID) to a DHCP server when acquiring a dynamic IPv6
+    address. DHCPv4 client protocol sends IAID and DUID to the DHCP server when acquiring
+    a dynamic IPv4 address if 'ClientIdentifier=duid'. IAID and DUID allows a DHCP server
```
Comment (keszybz) [inline man/networkd.conf.xml]: Please use `<option>ClientIdentifier=duid</option>` for markup.

File: man/networkd.conf.xml [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+    <para>The following options are available in <literal>[DUID]</literal> section:</para>
+
+    <variablelist class='network-directives'>
+
+      <varlistentry>
+        <term><varname>Type=</varname></term>
+        <listitem><para>The type of DUID specified in this section. The following values are
+        supported:</para>
+        <para>raw : If <literal>Type=raw</literal>, then <literal>RawData=</literal> specifies
```
Comment (keszybz) [inline man/networkd.conf.xml:98]: What about the other `Type=` settings that are accepted by `config_parse_duid_type`? Shouldn't they be documented here too?

Comment (vinaykul) [inline man/networkd.conf.xml:98]: Yes. I'll add them to the documentation marking them as a future support item so that the intent of this configuration interface is not lost in a issues thread / exists in just my or Tom's head. I'll update IAIDValue references to just IAID and squash the commits within the next day or two.

**Comment by keszybz** [comment]
OK, let's merge this. We usually don't describe things which _might_ be supported in the future, but the missing functionality should be relatively easy to fill in, hopefully we can do that soon.

There is some further work to do on the man page: the formatting is rather bad. But that's a minor thing, we can do it in subsequent commits.

**Comment by vinaykul** [comment]
@keszybz I just noticed the spurious newlines in the man pages display, not seen in the html version. Thanks for such a thorough review! I'll fix it in a future commit. Are there any other problems besides that?

I'm working on a patch for querying DUID and IAID values (Issue https://github.com/systemd/systemd/issues/2861) by adding a /run/systemd/netif/leases6/<ifindex> file to hold the DHCPv6 lease information, similar to /run/systemd/netif/leases/<ifindex> for IPv4, and extending networkctl status command to display CLIENTID (IPv4) and DHCP IAID and DUID for IPv6. Please let me know if you or @teg would approach this differently.

I hope to find some cycles to complete the DUID configuration task after this query work is done and a couple of other release critical tasks are squared away.

**Comment by keszybz** [comment]
If you get a chance to work on the man page, it would be nice to use proper markup for the term definitions. In particular, terms like "raw" should be marked up. See _TRANSPORT in systemd.journal-fields.7 for an example. This would make things consistent with other man pages and the terms would show up in systemd.directives.7. Also usually there's no space before a colon, that's only a French thing.

>  /run/systemd/netif/leases6/

Sounds good.

File: man/networkd.conf.xml [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+        <term><varname>RawData=</varname></term>
+        <listitem><para>Specifies the DUID bytes as a single newline-terminated, hexadecimal
+        string, with each byte separated by a ':'.</para></listitem>
+      </varlistentry>
+    </variablelist>
+  </refsect1>
+
+  <refsect1>
+    <para>The following options will be supported in the future:
+    </para>
```
Comment (poettering) [inline man/networkd.conf.xml:125]: This I really don't like. Either we implement something and document it, or we don't implement it and don't document it. But documenting behaviour of some potential future addition sounds really wrong, as only implementing something will allows to figure hout how the behaviour should actually be like.

Please, drop this, or implement this fully.

File: man/networkd.conf.xml [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+
+    <para>These configuration files control global network parameters.
+    For e.g. DHCP Unique Identifier (DUID).</para>
+
+  </refsect1>
+
+  <xi:include href="standard-conf.xml" xpointer="main-conf" />
+
+  <refsect1>
+    <title>[DUID] Section Options</title>
```
Comment (poettering) [inline man/networkd.conf.xml:68]: Hmm, I am pretty sure there should only be a [DHCP] section here, not a [DUID] section.

And the [DHCP] option should then have settings such as UniqueIdType=, UniqueIdRaw= and so on. While "DHCP" is a pretty widely known acronym, DUID really isn't. As such I'd prefer to have a more explanatory setting here, and thus call this UniqueIdXYZ= and similar.

If this stuff is placed in the [DHCP] section in this config file, this has the benefit that this maps nicely to .network files, where we can use the same optoins then, as the [DHCP] section already exists there.

File: src/network/networkd-conf.c [hunk 14]
```diff
@@ -0,0 +1,155 @@
+/*-*- Mode: C; c-basic-offset: 8; indent-tabs-mode: nil -*-*/
```
Comment (poettering) [inline src/network/networkd-conf.c:1]: we don't place these headers anymore in the sources

File: src/network/networkd-conf.c [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+        }
+
+        if (*duid_type != DUID_TYPE_RAW)
+                *dhcp_duid_type = (uint16_t)(*duid_type);
+
+        /* RawData contains DUID in format " NN:NN:NN... " */
+        while (true) {
+                r = extract_first_word(&pduid, &cbyte, ":", 0);
+                if (r < 0) {
+                        log_error("Failed to read DUID.");
```
Comment (poettering) [inline src/network/networkd-conf.c:124]: please use log_syntax() for logging configuration file parsing errors. See the other parser calls. Also, invalid input should generally not result in total failure, but just warnings, for compat reasons. Only things like OOM and so on should result in total failure.

File: src/network/networkd-conf.c [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+                /* First two bytes of RawData is DUID Type - fall thru */
+        default:
+                break;
+        }
+
+        if (*duid_type != DUID_TYPE_RAW)
+                *dhcp_duid_type = (uint16_t)(*duid_type);
+
+        /* RawData contains DUID in format " NN:NN:NN... " */
+        while (true) {
```
Comment (poettering) [inline src/network/networkd-conf.c:121]: Please use for(;;) for loops like this.

File: src/network/networkd-conf.c [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+                if (r == 0)
+                        break;
+                if (duid_index >= MAX_DUID_LEN) {
+                        log_error("DUID length exceeds maximum length.");
+                        return -EINVAL;
+                }
+
+                errno = 0;
+                byte = strtol(cbyte, &pnext, 16);
+                if ((errno == ERANGE && (byte == LONG_MAX || byte == LONG_MIN))
```
Comment (poettering) [inline src/network/networkd-conf.c:136]: I really don#t like the use of strtol() anywhere in our code. use unhexchar() or something like that.

Even better use unhexmem() and don't bother with the colons...

File: man/systemd.network.xml [hunk 18]
```diff
@@ -824,6 +830,86 @@
...
+    <para>The following options are available in <literal>[DUID]</literal> section:</para>
+
+    <variablelist class='network-directives'>
+
+      <varlistentry>
+        <term><varname>Type=</varname></term>
+        <listitem><para>The type of DUID specified in this section. The following values are
+        supported:</para>
+        <para>raw : If <literal>Type=raw</literal>, then <literal>RawData=</literal> specifies
+        the entire DUID. For e.g: <literal>RawData=00:02:00:00:ab:11:f9:2a:c2:77:29:f9:5c:00</literal>
```
Comment (poettering) [inline man/systemd.network.xml:860]: Is this colon-separated syntax generally accepted, or something made up here? If so, I'd prefer if we'd just take normal hex chars, without colons, the way we generate it otherwise with hexmem() or parse it with unhexmem().

File: src/network/networkd-conf.h [hunk 19]
```diff
@@ -0,0 +1,36 @@
+/*-*- Mode: C; c-basic-offset: 8; indent-tabs-mode: nil -*-*/
```
Comment (poettering) [inline src/network/networkd-conf.h:1]: please do not add this line to new sources.

**Comment by poettering** [comment]
I added a number of additional comments to the commits now. I can't say I like the way this is exposed currently.

File: man/networkd.conf.xml (hunk 12 above)
Comment (vinaykul) [inline man/networkd.conf.xml:125]: OK. I'll revert back to the previous version of the documentation, dropping the intended types for now... I might not get around to the extensions very soon.

File: man/networkd.conf.xml (hunk 13 above)
Comment (vinaykul) [inline man/networkd.conf.xml:68]: Tom suggested [DUID] - https://github.com/systemd/systemd/pull/2761#issuecomment-190835118. I think it helps avoid prefixing each key with DUID as originally discussed in https://github.com/systemd/systemd/issues/394#issuecomment-124225892 and consequently looks cleaner. 

@teg What are your thoughts about this?

File: man/systemd.network.xml (hunk 18 above)
Comment (vinaykul) [inline man/systemd.network.xml:860]: It is quite common to use ':' delimiter - e.g MAC address, IPv6 address, DUID displayed by dhcpcd, and in dhclient.conf . This is more readable, and if the user needs to enter it by hand vs. copy/paste, this is less error-prone.

**Comment by vinaykul** [comment]
@poettering @keszybz Please review PR https://github.com/systemd/systemd/pull/2943 that addresses the latest review items above. Waiting for @teg to weigh in on whether we want to keep the separate [DUID] section.
